src/algorithms/fardetectors/MatrixTransferStatic.cc

```diff
@@ -0,0 +1,154 @@
+    return;
+  }
+
+  aYinv[0][0] =  aY[1][1] / det;
+  aYinv[0][1] = -aY[0][1] / det;
+  aYinv[1][0] = -aY[1][0] / det;
+  aYinv[1][1] =  aY[0][0] / det;
+
+  return;
+
+}
+
+std::unique_ptr<edm4eic::ReconstructedParticleCollection> eicrecon::MatrixTransferStatic::produce(const edm4hep::SimTrackerHitCollection& rawhits) {
+
+  auto outputParticles = std::make_unique<edm4eic::ReconstructedParticleCollection>();
+
+  //---- begin Reconstruction code ----
+
+  double goodHitX[2] = {0.0, 0.0};
+  double goodHitY[2] = {0.0, 0.0};
+  double goodHitZ[2] = {0.0, 0.0};
+
+  bool goodHit1 = false;
+  bool goodHit2 = false;
+
+  for (const auto &h: rawhits) {
+
+    auto cellID = h.getCellID();
+    // The actual hit position in Global Coordinates
+    auto gpos = m_cellid_converter->position(cellID);
+    // local positions
+    auto volman = m_detector->volumeManager();
+    auto local = volman.lookupDetElement(cellID);
+
+    auto pos0 = local.nominal().worldToLocal(dd4hep::Position(gpos.x(), gpos.y(), gpos.z())); // hit position in local coordinates
+
+    if(!goodHit2 && gpos.z()/dd4hep::mm > m_cfg.hit2minZ && gpos.z()/dd4hep::mm < m_cfg.hit2maxZ){
+
+      goodHitX[1] = pos0.x()/dd4hep::mm;
+      goodHitY[1] = pos0.y()/dd4hep::mm;
+      goodHitZ[1] = gpos.z()/dd4hep::mm;
+      goodHit2 = true;
+
+    }
+    if(!goodHit1 && gpos.z()/dd4hep::mm > m_cfg.hit1minZ && gpos.z()/dd4hep::mm < m_cfg.hit1maxZ){
+
+      goodHitX[0] = pos0.x()/dd4hep::mm;
+      goodHitY[0] = pos0.y()/dd4hep::mm;
+      goodHitZ[0] = gpos.z()/dd4hep::mm;
+      goodHit1 = true;
+
+    }
+
+  }
+
+  // NB:
+  // This is a "dumb" algorithm - I am just checking the basic thing works with a simple single-proton test.
+  // Will need to update and modify for generic # of hits for more complicated final-states.
+
+  if (goodHit1 && goodHit2) {
+
+    // extract hit, subtract orbit offset – this is to get the hits in the coordinate system of the orbit
+    // trajectory
+    double XL[2] = {goodHitX[0] - m_cfg.local_x_offset, goodHitX[1] - m_cfg.local_x_offset};
+    double YL[2] = {goodHitY[0] - m_cfg.local_y_offset, goodHitY[1] - m_cfg.local_y_offset};
+
+    double base = goodHitZ[1] - goodHitZ[0];
+
+    if (base == 0) {
+      m_log->info("Detector separation = 0! Cannot calculate slope!");
+    }
+    else{
+
+      double Xip[2] = {0.0, 0.0};
+      double Xrp[2] = {XL[1], (1000 * (XL[1] - XL[0]) / (base)) - m_cfg.local_x_slope_offset}; //- _SX0RP_;
+      double Yip[2] = {0.0, 0.0};
+      double Yrp[2] = {YL[1], (1000 * (YL[1] - YL[0]) / (base)) - m_cfg.local_y_slope_offset}; //- _SY0RP_;
+
+      // use the hit information and calculated slope at the RP + the transfer matrix inverse to calculate the
+      // Polar Angle and deltaP at the IP
+
+      for (unsigned i0 = 0; i0 < 2; i0++) {
+	for (unsigned i1 = 0; i1 < 2; i1++) {
+	  Xip[i0] += aXinv[i0][i1] * Xrp[i1];
+	  Yip[i0] += aYinv[i0][i1] * Yrp[i1];
+	}
+      }
+
+      // convert polar angles to radians
```

Comment:
Avoid hard coding units.

Reply:
The matrix and offsets assume certain units are being used (because they are used to calculate the matrix in the first place).